src/digest/digest.cr

```diff
@@ -133,13 +133,68 @@ abstract class Digest
     final dst
   end
 
+  # Writes the final digest output to `dst`.
+  #
+  # Faster than the `Bytes` allocating version.
+  # Use when hashing in loops.
+  #
+  # This method can only be called once and raises `FinalizedError` on subsequent calls.
+  #
+  # NOTE: `.dup.final(dst)` call may be used to get an intermediate hash value.
   def final(dst : Bytes) : Bytes
     check_finished
     @finished = true
     final_impl dst
     dst
   end
 
+  # Returns a hexadecimal-encoded digest in a new `String`.
+  #
+  # This method can only be called once and raises `FinalizedError` on subsequent calls.
```

Comment:
The "can only be called once" is not limited to this method, but includes all finalizing methods. You can only call one of them, not each method once.

We could document that with references to `#finalize` in all methods' docs, or references to all methods from all methods.

Reply:
See latest "documentation" commit